libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp

```diff
@@ -0,0 +1,399 @@
+
+    int32_t vx = input_buffer[9] | (((uint32_t) input_buffer[10]) << 8)
+            | (((uint32_t) input_buffer[11]) << 16)
+            | (((uint32_t) input_buffer[12]) << 24);
+    hedge->positionBuffer[ind].x = vx;
+
+    int32_t vy = input_buffer[13] | (((uint32_t) input_buffer[14]) << 8)
+            | (((uint32_t) input_buffer[15]) << 16)
+            | (((uint32_t) input_buffer[16]) << 24);
+    hedge->positionBuffer[ind].y = vy;
+
+    int32_t vz = input_buffer[17] | (((uint32_t) input_buffer[18]) << 8)
+            | (((uint32_t) input_buffer[19]) << 16)
+            | (((uint32_t) input_buffer[20]) << 24);
+    hedge->positionBuffer[ind].z = vz;
+
+    hedge->positionBuffer[ind].highResolution = true;
+
+    ind = markPositionReady();
+
+    p = hedge->positionBuffer[ind];
+}
+
+uint16_t AP_Beacon_Marvelmind::CalcCrcModbus_(uint8_t * buf, int len) {
+    uint16_t crc = 0xFFFF;
+    int pos;
+    for (pos = 0; pos < len; pos++) {
+        crc ^= (uint16_t) buf[pos]; // XOR byte into least sig. byte of crc
+        int i;
+        for (i = 8; i != 0; i--) // Loop over each bit
+                {
+            if ((crc & 0x0001) != 0) // If the LSB is set
+                    {
+                crc >>= 1; // Shift right and XOR 0xA001
+                crc ^= 0xA001;
+            } else
+                // Else LSB is not set
+                crc >>= 1; // Just shift right
+        }
+    }
+    return crc;
+}
+
+void AP_Beacon_Marvelmind::createMarvelmindHedge() {
+
+    hedge->maxBufferedPositions = 3;
+    hedge->positionBuffer = NULL;
+    hedge->verbose = false;
+    hedge->receiveDataCallback = NULL;
+    hedge->lastValuesCount_ = 0;
+    hedge->lastValues_next = 0;
+    hedge->haveNewValues_ = false;
+    hedge->terminationRequired = false;
+
+}
+
+bool AP_Beacon_Marvelmind::healthy() {
+    // healthy if we have parsed a message within the past 300ms
+    return ((AP_HAL::millis() - last_update_ms) < AP_BEACON_TIMEOUT_MS);
+}
+
+void AP_Beacon_Marvelmind::startMarvelmindHedge() {
+    uint8_t i;
+    hedge->positionBuffer = (PositionValue*) malloc(
+            sizeof(struct PositionValue) * hedge->maxBufferedPositions);
+    if (hedge->positionBuffer == NULL) {
+        if (hedge->verbose)
+            puts("Not enough memory");
+        hedge->terminationRequired = true;
+        return;
+    }
+    for (i = 0; i < hedge->maxBufferedPositions; i++) {
+        hedge->positionBuffer[i].ready = false;
+        hedge->positionBuffer[i].processed = false;
+    }
+    hedge->positionsBeacons.numBeacons = 0;
+    hedge->positionsBeacons.updated = false;
+}
+
+void AP_Beacon_Marvelmind::update(void) {
+    static uint8_t counter = 0;
+    counter++;
+    if (counter > 200) {
+        counter = 0;
+    }
+
+    if (uart == nullptr) {
+        return;
+    }
+
+    // read any available characters
+    int32_t nBytesRead = uart->available();
+
+    uint8_t receivedChar = 0;
+    bool readSuccessed = true;
+    if (nBytesRead < 0)
```

Comment:
maybe simply with  if (nBytesRead < 0) { 
return; 
}
and get ride of readSuccessed, no ?